### PR TITLE
👓  phi3-vision tutorial

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes

 phi3-vision tutorial

image

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: examples/phi3/README_VISION.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Phi3 optimization with Olive
+This folder contains an example of optimizing the microsoft/Phi-3-vision-128k-instruct model from [Hugging Face](https://huggingface.co/microsoft/Phi-3-vision-128k-instruct) with Olive.
+
+## Prerequisites
+Please refer to the [pre-requisites of Phi3-vision optimization with GenAI](https://github.com/microsoft/onnxruntime-genai/blob/main/examples/python/phi-3-vision.md#0-pre-requisites) for more details.
+Basically, you need to install the following packages:
+- `huggingface_hub[cli]`
```
Comment (jambayk) [inline examples/phi3/README_VISION.md:7]: maybe the requirements other than ort-nightly can be moved into a requirement.txt file? Otherwise, the user needs to copy paste each one

File: examples/phi3/phi3_vision.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        default="cpu",
+        required=False,
+        choices=TARGETS,
+        help=f"Choose from {TARGETS}",
+    )
+    parser.add_argument(
+        "--precision",
+        type=str,
+        default="int4",
+        choices=["int4"],
```
Comment (jambayk) [inline examples/phi3/phi3_vision.py:38]: float models are not supported?

File: examples/phi3/vision/config_templates/vision_config.json [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            "input_types": [ "float32", "int64" ],
+            "input_shapes": [ [ 1, 17, 3, 336, 336 ], [ 1008, 1344 ] ],
+            "output_names": [ "visual_features" ],
+            "dynamic_axes": {
+                "pixel_values": { "0": "num_images", "1": "max_num_crops", "3": "height", "4": "width" },
+                "image_sizes": { "0": "num_images" },
+                "visual_features": { "0": "batch_size", "1": "num_img_tokens" }
+            }
+        },
+        "dummy_inputs_func": "get_dummy_inputs"
```
Comment (jambayk) [inline examples/phi3/vision/config_templates/vision_config.json:18]: Is the dummy input funce needed since we have all of the input shapes and types already in the ioconfig?

File: examples/phi3/phi3_vision.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    import numpy as np
+    import torch
+    from onnx import TensorProto, external_data_helper, helper, numpy_helper, save_model
+    from transformers import AutoConfig, AutoModelForCausalLM
+
+    model = AutoModelForCausalLM.from_pretrained(input_model_path, trust_remote_code=True)
+    config = AutoConfig.from_pretrained(input_model_path, trust_remote_code=True)
+
+    # User inputs
+    io_dtype = TensorProto.FLOAT16 if args.precision == torch.float16 else TensorProto.FLOAT
```
Comment (jambayk) [inline examples/phi3/phi3_vision.py:97]: args.precision appears to be a string with only "int4" as option. when would this case be true? I think we expect to use fp16 if the target is cuda and fp32 when it is cpu?

File: examples/phi3/vision/config_templates/vision_config.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
+{
+    "input_model": {
+        "type": "PyTorchModel",
+        "model_path": "<<str: model_path>>",
+        "model_loader": "vision_embed_tokens_loader",
+        "model_script": "scripts/user_script.py",
```
Comment (jambayk) [inline examples/phi3/vision/config_templates/vision_config.json:6]: this path is relative to the "vision" folder but we are running from "phi3"

File: examples/phi3/phi3_vision.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+    else:
+        config["passes"]["convert"]["torch_dtype"] = "float16"
+        config["passes"]["convert"]["device"] = "cuda"
+        config["systems"]["local_system"]["accelerators"] = [
+            {"device": "GPU", "execution_providers": ["CUDAExecutionProvider"]}
+        ]
+
+    config["engine"] = {
+        "cache_dir": Path(args.cache_dir).resolve() / "vision",
+        "output_dir": Path(args.output_dir).resolve() / "vision",
```
Comment (jambayk) [inline examples/phi3/phi3_vision.py:283]: the output of the config run still stays in the output directory. It should probably be cleaned up after the resave

File: examples/phi3/phi3_vision.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+        processor = AutoProcessor.from_pretrained(input_model_path, trust_remote_code=True)
+        model = AutoModelForCausalLM.from_pretrained(input_model_path, trust_remote_code=True)
+        del model, processor, config
+    except Exception as e:
+        print(f"Error loading model: {e}")
+        return False
+    return True
+
+
+def build_text_embedding(args, input_model_path):
```
Comment (jambayk) [inline examples/phi3/phi3_vision.py:84]: Is this manual building of the model necessary? I am able to generate the same graph using onnx conversion of the embed_tokens submodule https://github.com/microsoft/Olive/commit/777d82250e8a858b1fe584aa90e6b5b56164b478

File: examples/phi3/README_VISION.md (hunk 1 above)
Comment (trajepl) [reply examples/phi3/README_VISION.md]: Updated.

File: examples/phi3/phi3_vision.py (hunk 2 above)
Comment (trajepl) [reply examples/phi3/phi3_vision.py]: Add fp16 support. The fp32 of onnx model run failed, 

[E:onnxruntime:onnxruntime-genai, sequential_executor.cc:516 ExecuteKernel] Non-zero status code returned while running Expand node. Name:'/model/attn_mask_reformat/input_ids_subgraph/Expand' Status Message: /model/attn_mask_reformat/input_ids_subgraph/Expand: left operand cannot broadcast on dim 3 LeftShape: {1,1,2500,2500}, RightShape: {1,1,2500,5572}

File: examples/phi3/phi3_vision.py (hunk 4 above)
Comment (trajepl) [reply examples/phi3/phi3_vision.py]: Updated.

File: examples/phi3/vision/config_templates/vision_config.json (hunk 5 above)
Comment (trajepl) [reply examples/phi3/vision/config_templates/vision_config.json]: The value will be updated in phi3-vision.py. To avoid confusion, updated.

File: examples/phi3/phi3_vision.py (hunk 6 above)
Comment (trajepl) [reply examples/phi3/phi3_vision.py:297]: The output_dir can be changed to another one. Since the model name must follow the given format to aligh the that in generation_config.json.

File: examples/phi3/phi3_vision.py (hunk 7 above)
Comment (trajepl) [reply examples/phi3/phi3_vision.py:88]: https://huggingface.co/microsoft/Phi-3-vision-128k-instruct-onnx-cpu/blob/main/onnx/builder.py
It is borrowed from huggingface builder.py which is mentined in genai's tutorial as well.
I am not sure if it is totally same with build_text_embedding.

Comment (jambayk) [reply examples/phi3/phi3_vision.py:88]: Since the graph looks the same, could you try the model generated from the config. 

It feels unnecessary and non standard to manually generate the graph is a simple exported model works.

File: examples/phi3/phi3_vision.py [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+        save_as_external_data=pass_config.get("save_as_external_data", False),
+        all_tensors_to_one_file=pass_config.get("all_tensors_to_one_file", True),
+        external_data_name=pass_config.get("external_data_name", None),
+        size_threshold=pass_config.get("size_threshold", 1024),
+        convert_attribute=pass_config.get("convert_attribute", False),
+    )
+
+
+def main(raw_args=None):
+    args = get_args(raw_args)
```
Comment (jambayk) [inline examples/phi3/phi3_vision.py:201]: maybe you can check that fp16 is not allowed for cpu target?

File: examples/phi3/phi3_vision.py [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+        pass_config = output_node.pass_run_config
+        resave_onnx_model(model_path, output_dir / "phi-3-v-128k-instruct-text.onnx", pass_config)
+    except Exception:
+        config_link = "https://huggingface.co/microsoft/Phi-3-vision-128k-instruct-onnx-cpu/tree/main/cpu-int4-rtn-block-32-acc-level-4."
+        print(
+            "Ignore the error during Olive run.\nThis script will copy the "
+            "genai_config.json and processor_config.json from ",
+            config_link,
+        )
+        # even if the Olive run fails, partially generated model
```
Comment (jambayk) [inline examples/phi3/phi3_vision.py:260]: Do you know what the failure could be? Does it fail during quantization?

File: examples/phi3/phi3_vision.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    config["input_model"]["model_path"] = input_model_path
+    config["input_model"]["model_script"] = config_path.parent / "scripts" / "user_script.py"
+
+    if args.precision == "fp16" or (args.precision == "int4" and args.target == "cuda"):
+        config["passes"]["convert"]["torch_dtype"] = "float16"
+    else:
+        config["passes"]["convert"]["torch_dtype"] = "float32"
+
+    if args.target == "cpu":
+        config["passes"]["convert"]["torch_dtype"] = "float32"
```
Comment (jambayk) [inline examples/phi3/phi3_vision.py:285]: nit: this looks redundant? Since the float16 case should only happen for cpu if we enforce the check earlier.

File: examples/phi3/phi3_vision.py (hunk 6 above)
Comment (jambayk) [reply examples/phi3/phi3_vision.py:297]: Since the workflow output_dir (not the output dir of this script) contents are already resaved into the final directory, I think this still needs to be cleaned up regardless of where it is. 

Otherwise, we just have copies of the same model lying around.

File: examples/phi3/phi3_vision.py (hunk 9 above)
Comment (trajepl) [reply examples/phi3/phi3_vision.py:260]: Failure in model builder. They had a fix in main branch already. After the new version of genai is released, this failure will be resolved.

Comment (trajepl) [reply examples/phi3/phi3_vision.py:260]: https://github.com/microsoft/onnxruntime-genai/blob/cb313b49adec51f8679701df7d2451c9d21374a2/src/python/py/models/builder.py#L267C1-L269C129

File: examples/phi3/phi3_vision.py (hunk 7 above)
Comment (trajepl) [reply examples/phi3/phi3_vision.py:89]: image
Conversion cannot not work directly. Seems the generator requires the inputs name of `inputs_embeds`.

Comment (trajepl) [reply examples/phi3/phi3_vision.py]: It worked. Just updated.

File: examples/phi3/phi3_vision.py (hunk 6 above)
Comment (trajepl) [reply examples/phi3/phi3_vision.py:213]: Removed.

File: examples/phi3/phi3_vision.py [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+        cache_dir = text_config.engine.cache_dir
+        pass_config = text_config.passes["builder"].config
+        # find the model file in the cache directory
+        model_path = list(cache_dir.rglob("*.onnx"))[-1]
+        resave_onnx_model(model_path, output_dir / "phi-3-v-128k-instruct-text.onnx", pass_config)
+    print("Model generation completed, output saved to ", output_dir)
+    # clean up the output directory in olive.workflows
+    to_remove_folders = [
+        Path(args.output_dir).resolve() / "vision",
+        Path(args.output_dir).resolve() / "text",
```
Comment (jambayk) [inline examples/phi3/phi3_vision.py:182]: text_embedding is missing here.

Otherwise, everything else looks good to me!

Comment (trajepl) [reply examples/phi3/phi3_vision.py:182]: Updated.

**Review by jambayk — APPROVED** [review]
Thanks!
